Add Algorithmia integration

koverholt opened the pull request:
### What does this PR do?

Add information about Algorithmia integration.

### Motivation

To add a tile and documentation with steps on integrating Algorithmia with Datadog.

### Review checklist

- [X] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [X] Feature or bugfix has tests
- [X] Git history is clean
- [X] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

N/A

irabinovitch commented on algorithmia/metadata.csv:
```diff
@@ -0,0 +1,3 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+duration_milliseconds,gauge,,millisecond,,Duration of algorithm run,0,algorithmia,
+<USER-DEFINED>,gauge,,,,Inference-related metrics defined by an algorithm developer,0,algorithmia,
```
We'll need this to be a specific / consistently named set of metrics.

koverholt replied:
That makes sense, thanks! I removed the `<USER-DEFINED>` row from `metadata.csv` and instead clarified the point about additional user-defined metrics in the README.

irabinovitch commented on algorithmia/metadata.csv:
```diff
@@ -0,0 +1,2 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+duration_milliseconds,gauge,,millisecond,,Duration of algorithm run,0,algorithmia,
```
I think we're still missing a metric name here. Is it just the 1 metric?

koverholt replied:
Good question! I can give you some context, and perhaps you could give me some guidance as to whether I should list some other values here.

A example of a payload from Algorithmia to the Datadog metrics API looks like:

```
{
  "time": "2020-09-29T05:02:13.546283",
  "duration_milliseconds": 16,
  "session_id": "rses-981e1fbf-e3a6-44ae-ad33-cc00599bc687",
  "request_id": "req-d717d20b-49e9-466e-a370-1a6cf405486b",
  "algorithm_owner": "koverholt",
  "algorithm_name": "model",
  "algorithm_version": "1.0.0",
  "<USER-DEFINED-METRIC-1>": 4,
  "<USER-DEFINED-METRIC-2>": 0.98,
  "<USER-DEFINED-METRIC-3>": 1,
  "<USER-DEFINED-METRIC-N>": 0.57
}
```

where `<USER-DEFINED-METRIC-N>` are defined by the user in our platform but are not present by default, `duration_milliseconds` is currently listed as the only metric in `metadata.csv`, and all of the other key/values would be used as facets in Datadog rather than metrics.

So, given that information, should I include the keys/values in `metadata.csv` that would be considered to be facets in Datadog? If so, then I can add them. If not, then the only pre-defined metric from a default payload would be `duration_milliseconds` as currently shown in `metadata.csv`.

irabinovitch replied:
metatdata.csv should only include metrics with stable/known names with consistent units.

hithwen commented on .github/CODEOWNERS:
```diff
@@ -1,5 +1,6 @@
 *                       @DataDog/agent-integrations
 /*/README.md            @DataDog/documentation
+/algorithmia/           support@algorithmia.io
```
can you please add a github username or team here as well?

koverholt replied:
Done. Added a Github username there. 👍

hithwen commented on algorithmia/metadata.csv:
```diff
@@ -0,0 +1,2 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+duration_milliseconds,gauge,,millisecond,,Duration of algorithm run,0,algorithmia,
```
As it says in https://github.com/DataDog/integrations-extras/pull/727/files#diff-7eac1194a680f15116e7155e23416f62R6 metrics emitted by this integration should be prefixed with `algorithmia.` to be correctly identified as belonging to this integration. Although here https://github.com/DataDog/integrations-extras/pull/727/files#diff-a46bf5e60a1167da9b93115a5fae52a5R55-R56 it says the prefix is `algorithmia-insights.`.
Whichever is the correct prefix it should be reflected both in the `metadata.csv` and the `manifest.json` files (and also it must be true that the metrics are coming with that prefix)

koverholt replied:
Fixed. Made the prefix consistent between `metadata.csv`, `manifest.json`, and the README. Thanks!

kayayarai commented on algorithmia/README.md:
```diff
@@ -0,0 +1,89 @@
+Refer to [metadata.csv][5] for a list of operational metrics provided by this
+check. Additional metrics can include any user-defined, inference-related
+metrics that are specified by the algorithm developer.
+
```
```suggestion

See [metadata.csv][5] for a list of metrics provided by this integration.

```
This line magically embeds a table of the metrics in the documentation, so I'm adding it here below your desciption.

koverholt replied:
Nice, thank you! I fixed the wording in the README to use this sentence instead in d223d1d.

kayayarai commented on algorithmia/README.md:
```diff
@@ -0,0 +1,89 @@
+Insights enabled is queried. Each log entry includes operational metrics and
+inference-related metrics.
+
+Refer to [metadata.csv][5] for a list of operational metrics provided by this
```
```suggestion
Refer to [metadata.csv][5] or the table below for a list of operational metrics provided by this
```

kayayarai reviewed (changes requested):
Really solid writing in the Readme. Thank you for this! I added a sentence that our build uses to pull in a table of metrics right into the docs, and changed the wording around that a bit.

koverholt commented:
Thank you all for the review comments! I've fixed all of the issues in the most recent commits. Let me know if I can address anything else.

hithwen commented on algorithmia/manifest.json:
```diff
@@ -0,0 +1,27 @@
+    "saved_views": {},
+    "monitors": {},
+    "service_checks": "assets/service_checks.json",
+    "logs": {}
```
```suggestion
    "logs": {},
    "metrics_metadata": "metadata.csv"
```
This is a new field we've just introduced

koverholt replied:
Added, thanks!

kayayarai reviewed (approved):
readme updates look good

ofek reviewed (approved):
@koverholt Great work, thanks!!!

koverholt commented:
Hi All:

Thanks for the reviews and feedback so far! Based on our ongoing partner conversations, I reworked how our integration flows data from Kafka to Datadog.

The initial integration was using the a Kafka Connector that was sending our metrics to the Datadog logs API.

The updated integration that I just added is a standalone Python service that forwards our metrics to the Datadog metrics API. I've also included an out-of-the-box dashboard and monitor that is configured for our default metric and described how users can add user-defined metrics that they add from Algorithmia.

I believe that I updated all of the relevant documentation and other components in this PR, please let me know if there is anything else that I need to update. Thanks!

ChristineTChen commented on algorithmia/manifest.json:
```diff
@@ -0,0 +1,28 @@
+  "manifest_version": "1.0.0",
+  "name": "algorithmia",
+  "metric_prefix": "algorithmia.",
+  "metric_to_check": "",
```
```suggestion
  "metric_to_check": "algorithmia.duration_milliseconds",
```
We should specify a metric here to ensure that the tile gets automatically installed when this metric is detected.

koverholt replied:
Done in d81ad5f, thanks!

ChristineTChen commented on algorithmia/manifest.json:
```diff
@@ -0,0 +1,28 @@
+  "is_public": false,
+  "integration_id": "algorithmia",
+  "assets": {
+    "dashboards": {},
```
I see you've added a dashboard but it's not linked here. Did you use the `ddev meta dash export command` to properly add the dashboard to the integration? See more on the command + what it does here: https://datadoghq.dev/integrations-core/guidelines/dashboards/#exporting-a-dashboard-payload

alexvetras replied:
The link Christine provided offers some examples of best practices for dashboards. 

In particular, could you please add the following:

1. Algorithmia's Logo
2. A description of the dashboard to educate the user of its purpose/value (you can also add links to any Algorithmia documentation, a blogpost, Github, etc.)
3. Clear grouping titles, and graph titles, along with descriptions on the side explaining what the graph shows.

Similarly for the monitor, please ensure a thoughtful threshold value, add a description, and include an action playbook on how someone receiving the alert should handle it. 

Thanks!

koverholt replied:
I added the dashboard and monitor to the manifest manually in 6e41962 since I was getting an error when using the `ddev meta dash export` command.

koverholt replied:
I think that I followed all of the right steps, the error that I am getting is:

```
Traceback (most recent call last):
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/toml/decoder.py", line 511, in loads
    multibackslash)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/toml/decoder.py", line 736, in load_line
    raise ValueError("Invalid date or number")
ValueError: Invalid date or number

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/koverholt/miniconda3/bin/ddev", line 8, in <module>
    sys.exit(ddev())
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/click/core.py", line 1256, in invoke
    Command.invoke(self, ctx)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/datadog_checks/dev/tooling/cli.py", line 43, in ddev
    config = load_config()
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/datadog_checks/dev/tooling/config.py", line 79, in load_config
    config.update(toml.loads(read_config_file()))
  File "/Users/koverholt/miniconda3/lib/python3.7/site-packages/toml/decoder.py", line 513, in loads
    raise TomlDecodeError(str(err), original, pos)
toml.decoder.TomlDecodeError: Invalid date or number (line 38 column 1 char 562)
```

Let me know if I missed anything by manually adding the dashboard and monitor.

koverholt replied:
Also, thanks for the tips, @alexvetras. Let me get my `ddev` back in working order and I will add the changes that you mentioned regarding the dashboard and monitor best practices.

koverholt replied:
I updated the dashboard and monitor in dc38a9f and 1fbdc3d to follow the documented best practices.

<img width="1511" alt="Screen Shot 2020-10-22 at 10 58 45 PM" src="https://user-images.githubusercontent.com/303236/96954989-345e4d00-14ba-11eb-98b0-0fe89b3c7cf7.png">

* Use screenboard instead of timeboard
* Insert logo and descriptive notes
* Add explanations of default operational metrics and how to add user-specified metrics
* Added monitor threshold, description, `recommended_monitor_metadata`, and steps to take upon an alert

Thanks again for the guidance here!

ChristineTChen commented on algorithmia/src/kafka-datadog.py:
```diff
@@ -0,0 +1,61 @@
+from datadog import initialize, api
+from kafka import KafkaConsumer
+
+try:
```
Just a question, is there any reason we don't want to this to be a check that runs with an agent?

alexvetras replied:
Adding Michael's explanation here based on our call with Kris today. 

Algorithmia would like to read from the kafka streams directly which is a firehose of data and isn't well suited for a periodic check, hence this direct to API approach. We highlighted the alternative to use DogStatsD with a local Agent install to get some of the agent benefits, but that will likely be a future evolution for Algorithmia. 

Thanks!

koverholt replied:
Thanks, @alexvetras, that is correct. I prototyped an approach using an agent check before submitting this version, and because hooking into a Kafka pipeline is a blocking call rather than a finite check, it hangs when the agent check runs, hence the approach used here (and the plans for future work that Alex mentioned for sending this data to DogStatsD).

hithwen replied:
If this is not an agent check then the python part should not be in this repo (owned by algorithmia?). That would also have the advantage for the users of not having to clone this entire repo to run it and for you to not depend on code reviews from Datadog's side

koverholt replied:
As we continue to work with customers on this integration, it is possible that this evolves into a script with custom metrics getting sent into DogStatsD. Also, having the Python message forwarding component here ensures a tighter coupling between the payload from Algorithmia and the default metrics that are configured in the dashboard and monitor in terms of the prefix, tags, etc. So, I am okay leaving the message forwarding component in this repo in the meantime if that is acceptable.

hithwen replied:
>  it is possible that this evolves into a script with custom metrics getting sent into DogStatsD

I understand that a DogStatsD emitter would also be an external process running separately from the agent so it would still not be an agent integration.
Hosting the code on this repo comes with following advantages, none of those apply to a separate process:
* CI set up (this has no tests)
* Validation to ensure it follows guidelines for developing agent checks (this is not a custom check)
* Wheel releases so integrations can be installed directly from the agent using a command (it would not apply since its not an agent check)
With the disadvantage of not having merge ownership

> Also, having the Python message forwarding component here ensures a tighter coupling between the payload from Algorithmia and the default metrics that are configured in the dashboard and monitor in terms of the prefix, tags, etc

That coupling is virtual because there's no automation that ensures it (versus on agent checks where automated testing can be added to ensure metadata.csv matches the actual metrics etc).

All the other datadog or dogstastd emitters that have integration tiles host the code themselves (and it's often part of the product to monitor itself rather than a separate script).

koverholt replied:
Got it, thanks for clarifying! Should I go ahead and move the Python component out of this repo and update the documentation here?

hithwen replied:
Yes, please

koverholt replied:
Thanks! In 7604a17, I moved the Python script and specific steps to our integrations repo at https://github.com/algorithmiaio/integrations/tree/master/Datadog, and I updated the documentation here to reflect the update.

ChristineTChen reviewed (changes requested):
Left some comments for the manifest.json to ensure the dashboard gets added

alexvetras reviewed (approved):
This dashboards looks good to me. Thanks for updating it, Kris.

alexvetras reviewed (approved):
These updates to the monitor also look to be in line with our best practices. Thanks!

apigirl commented on algorithmia/README.md:
```diff
@@ -0,0 +1,79 @@
+
+### Metrics
+
+This integration streams metrics from Algorithmia when an model that has
```
```suggestion
This integration streams metrics from Algorithmia when a model that has
```

koverholt replied:
Thanks for catching this, fixed!
